### PR TITLE
New upload manager

### PR DESCRIPTION
**Pull request by ezaquarii**
@tobiasKaminsky @AndyScherzinger 

This is the (too) long postponed effort to finish the upload manager, so far sabotaged by more or less fun life events. 

The implementation will build on top of existing downloader by extending it to handle bi-directional transfers.
There are still many white spots but the sneak preview in ETM using test mode is there.

There are still many issues to iron out, especially around various flaws of `OCFile` and `OCUpload` abstractions, but I'd welcome any preliminary feedback, feature requests, expectations, etc.

I plan to test it first with contacts uploader, to minimize the fallout.

### CODE REVIEW

**Comment by ezaquarii** [comment]
@tobiasKaminsky Could you please shed some light on the purpose of local behaviour options?

https://github.com/nextcloud/android/blob/109487ff915273819399fd562cc1823eb63fe3c2/src/main/java/com/owncloud/android/files/services/FileUploader.java#L175-L178

I understand that they determine what happens after upload completion, but I'm trying to understand what exactly they mean from UX perspective.

**Comment by tobiasKaminsky** [comment]
> @tobiasKaminsky Could you please shed some light on the purpose of local behaviour options?
> 
> https://github.com/nextcloud/android/blob/109487ff915273819399fd562cc1823eb63fe3c2/src/main/java/com/owncloud/android/files/services/FileUploader.java#L175-L178

UX wise you can find them when manually uploading a file:



Or when configuring auto upload:

**Comment by ezaquarii** [comment]
@tobiasKaminsky the errors seems to be flaky tests; I'd review the code nevertheless.

The API presentation is in https://github.com/nextcloud/android/blob/b3896ac9036b9d4fbb22dccdecc8103fec81a793/src/main/java/com/nextcloud/client/jobs/ContactsBackupWork.kt#L165-L176.

**Comment by ezaquarii** [comment]
@tobiasKaminsky any thoughts?

**Comment by ezaquarii** [comment]
@tobiasKaminsky Any chance we can get back to this?

**Comment by ezaquarii** [comment]
```
com.******.client.jobs.ContactsBackupIT > importExport[android(AVD) - 8.1.0] FAILED
	java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.owncloud.android.datamodel.OCFile.getRemotePath()' on a null object reference
	at com.owncloud.android.operations.RefreshFolderOperation.run(RefreshFolderOperation.java:229)
```

@tobiasKaminsky 
Not sure if that's me, but I doubt so as the uploader is not used anywhere except ETM test screen atm. Does it happen elsewhere?

**Comment by ezaquarii** [comment]
@tobiasKaminsky ping

File: src/main/java/com/nextcloud/client/files/downloader/TransferManagerImpl.kt [hunk 1]
```diff
@@ -144,4 +171,23 @@ class TransferManagerImpl(
         }
         return DownloadTask.Result(file, true)
     }
+
+    /**
+     *  Test upload task is used only to simulate upload process without
```
Comment (tobiasKaminsky) [inline src/main/java/com/nextcloud/client/files/downloader/TransferManagerImpl.kt:176]: Is this needed in our productive code or can we move this to AndroidTest?

**Comment by tobiasKaminsky** [comment]
Looks great! 
I am really really sorry for coming back this late to you :-(

I have only one minor question and then we can release it.

**Review by tobiasKaminsky — APPROVED** [review]

File: src/main/java/com/nextcloud/client/files/downloader/TransferManagerImpl.kt (hunk 1 above)
Comment (tobiasKaminsky) [reply src/main/java/com/nextcloud/client/files/downloader/TransferManagerImpl.kt:176]: Ah, this is while using in ETM.
So I am totally fine :-)

**Comment by tobiasKaminsky** [comment]
I'll give CI one last try, and then merge it :-)

File: src/main/java/com/nextcloud/client/files/downloader/TransferManagerImpl.kt (hunk 1 above)
Comment (ezaquarii) [reply src/main/java/com/nextcloud/client/files/downloader/TransferManagerImpl.kt:176]: At least for now. I'm fine removing it or moving into dev-only release when we finish workin on it.

**Comment by ezaquarii** [comment]
@tobiasKaminsky Planing to merge it? It's marked "to release" but I see no actions taken for few months now.

**Comment by ezaquarii** [comment]
I'm not sure what is happening here, but it doesn't look like a source code problem.
```
 * What went wrong:
Execution failed for task ':kaptGenericDebugKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.gradle.internal.KaptExecution
   > java.lang.reflect.InvocationTargetException (no error message)
```

**Pull request merged**